File: 17. Letter Combinations of a Phone Number.java
class Solution {
    public char[] numToLetter(int num){
        if(num == 2) return new char[]{'a', 'b', 'c'};
        else if(num == 3) return new char[]{'d', 'e', 'f'};
        else if(num == 4) return new char[]{'g', 'h', 'i'};
        else if(num == 5) return new char[]{'j', 'k', 'l'};
        else if(num == 6) return new char[]{'m', 'n', 'o'};
        else if(num == 7) return new char[]{'p', 'q', 'r', 's'};
        else if(num == 8) return new char[]{'t', 'u', 'v'};
        else return new char[]{'w', 'x', 'y', 'z'};
    }
    public void backtrack(List<String> ans, StringBuilder curr, String digits, int index){
        if(index >= digits.length()){
            ans.add(curr.toString());
            return;
        }

        int digit = digits.charAt(index) - '0';
        for(char ch : numToLetter(digit)){
            curr.append(ch);
            backtrack(ans, curr, digits, index + 1);
            curr.deleteCharAt(curr.length() - 1);
        }
    }
    public List<String> letterCombinations(String digits) {
        List<String> ans = new ArrayList<>();
        backtrack(ans, new StringBuilder(), digits, 0);
        
        return ans;
    }
}

File: 17. Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<char> numToLetter(int num) {
        if (num == 2) return {'a', 'b', 'c'};
        else if (num == 3) return {'d', 'e', 'f'};
        else if (num == 4) return {'g', 'h', 'i'};
        else if (num == 5) return {'j', 'k', 'l'};
        else if (num == 6) return {'m', 'n', 'o'};
        else if (num == 7) return {'p', 'q', 'r', 's'};
        else if (num == 8) return {'t', 'u', 'v'};
        else return {'w', 'x', 'y', 'z'};
    }

    void backtrack(vector<string>& ans, string& curr, string& digits, int index) {
        if (index >= (int)digits.size()) {
            ans.push_back(curr);
            return;
        }

        int digit = digits[index] - '0';
        for (char ch : numToLetter(digit)) {
            curr.push_back(ch);
            backtrack(ans, curr, digits, index + 1);
            curr.pop_back();
        }
    }

    vector<string> letterCombinations(string digits) {
        vector<string> ans;
        if (digits.empty()) return ans;
        string curr;
        backtrack(ans, curr, digits, 0);
        return ans;
    }
};
